core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -2136,7 +2138,10 @@ class SparkContext(config: SparkConf) extends Logging {
     }
     if (env != null) {
       Utils.tryLogNonFatalError {
-        env.metricsSystem.report()
+        val metricsReportFuture = Future {
+          env.metricsSystem.report()
+        } (SparkContext.executionContext)
+        ThreadUtils.awaitResult(metricsReportFuture, Duration.create("1s"))
```

Comment:
1s seems to be a bit aggressive. How long would it take normally?

Reply:
I think 1 to 2 seconds is reasonable.  The last metrics report is not necessary, we must reserve enough time for the application master hook to unregister and clean up temporary files.

What do you think?